File: bug_04_iterators/ts_code/iterators_01.ts
function get_strings() : string[] {
  return [
    "EEE", "DDD", "", "AAA", "CCC", "", "BBB", "FFF"
  ];
}

let make_lowercase = (s: string) : string => {
  return s.toLowerCase();
}

let vs = get_strings();
console.log("Unsorted strings:");
for (let s of vs) {
  console.log(s)
}
console.log("One more time to make sure all the bits are in the right places:", vs)

console.log("Sorted non-empty strings")
vs = vs.filter( str => str.length > 0 )
           .sort();
for (let s of vs) {
  console.log(s)
}

console.log("Lowercased sorted non-empty strings")
vs.forEach((s, index) => {
  vs[index] = make_lowercase(s)
});

console.log(vs);

File: bug_04_iterators/src/suggestion/iterators_01.rs

fn get_strings() -> Vec<String> {
    //map over the string slice array to transform items to
    // Strings and convert the array to Vec<String>
    [
        "EEE", "DDD", "", "AAA", "CCC", "", "BBB", "FFF"
    ]
    .map(|a| String::from(a))
    .to_vec()
}

//Borrow an mutable exclusive reference to the string in order
// to change it
fn make_lowercase(str: &mut String) {
    str.to_uppercase();
}

fn main() {
    let vs = get_strings();
    println!("Unsorted strings:");
    //Iterate over a shared reference to vs to avoid moving it
    // into the loop, thus consuming it
    let _vsr = &vs; //rust-analyzer shows the value type of the variable
    for s in &vs {// for iterators, rust-analyzer shows the value type of the items
        println!("{}", s);
    }

    println!("One more time to make sure all the bits are in the right places: {:?}", vs);

    //Filter the empty string items
    let mut vs: Vec<String> = vs.into_iter()
        .filter(|str| !str.is_empty())
        .collect();
    vs.sort();

    println!("Sorted non-empty strings: {:?}", vs);

    fn make_lowercase(s: &mut String) {
        *s = s.to_lowercase()
    }
    for s in vs.iter_mut() {
        make_lowercase(s);
    };
    println!("Lowercased strings: {:?}", vs);
}
